### PR TITLE
Runtime preparser options

### PR DESCRIPTION
**Pull request by pontusmelke**
Usage:

```
RUNTIME COMPILED MATCH n RETURN n
RUNTIME INTERPRETED MATCH n RETURN n
```

Default is `INTERPRETED`.

Naming is not yet final.

### CODE REVIEW

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/AmendedRootPlanDescription.scala [hunk 1]
```diff
@@ -54,13 +55,14 @@ class AmendedRootPlanDescription(inner: ExtendedPlanDescription, version: Cypher
...
-  override def toString = {
+  override def toString: String = {
     // Have to hack toString here as the alternative would be to release 1.9, 2.0, .. to add a new argument type
     val innerToString = childAsJava.toString
     val arguments = asJava.getArguments
     val version = arguments.get("version")
     val planner = arguments.get("planner")
-    s"Compiler $version \n\n$planner\n\n$innerToString"
+    val runtime = arguments.get("runtime")
+    s"Compiler $version \n\n$planner\n\n $planner\n\n$innerToString"
```
Comment (Mats-SX) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/AmendedRootPlanDescription.scala]: Second $planner should probably be $runtime

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/CompatibilityFor2_3.scala [hunk 2]
```diff
@@ -195,17 +195,17 @@ case class ExecutionResultWrapperFor2_3(inner: InternalExecutionResult, planner:
     new ResourceIterator[util.Map[String, Any]] {
       def close() = exceptionHandlerFor2_3.runSafely {
         endQueryExecution()
-        innerJavaIterator.close
+        innerJavaIterator.close()
```
Comment (Mats-SX) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/CompatibilityFor2_3.scala:198]: Why is this change needed?

Comment (pontusmelke) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/CompatibilityFor2_3.scala:198]: It is not needed per se, but if you declare a def with parenthesis you should also call it using parenthesis, it works without but you get compiler warnings.

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/AmendedRootPlanDescription.scala (hunk 1 above)
Comment (pontusmelke) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/AmendedRootPlanDescription.scala]: you're right, nice catch!
